pkg/binlog/reader/tcp.go

```diff
@@ -0,0 +1,159 @@
+	streamer  *replication.BinlogStreamer
+}
+
+// TCPReaderStatus represents the status of a TCPReader.
+type TCPReaderStatus struct {
+	Stage  string `json:"stage"`
+	ConnID uint32 `json:"connection"`
+}
+
+// String implements Stringer.String.
+func (s *TCPReaderStatus) String() string {
+	data, err := json.Marshal(s)
+	if err != nil {
+		log.Errorf("[TCPReaderStatus] marshal status to json error %v", err)
+	}
+	return string(data)
+}
+
+// NewTCPReader creates a TCPReader instance.
+func NewTCPReader(syncerCfg replication.BinlogSyncerConfig) Reader {
+	return &TCPReader{
+		syncerCfg: syncerCfg,
+		syncer:    replication.NewBinlogSyncer(syncerCfg),
+	}
+}
+
+// StartSyncByPos implements Reader.StartSyncByPos.
+func (r *TCPReader) StartSyncByPos(pos gmysql.Position) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if r.stage.Get() != int32(stageNew) {
+		return errors.Errorf("stage %s, expect %s", readerStage(r.stage), stageNew)
+	}
+
+	streamer, err := r.syncer.StartSync(pos)
+	if err != nil {
+		return errors.Annotatef(err, "start sync from position %s", pos)
+	}
+
+	r.streamer = streamer
+	r.stage.Set(int32(stagePrepared))
+	return nil
+}
+
+// StartSyncByGTID implements Reader.StartSyncByGTID.
+func (r *TCPReader) StartSyncByGTID(gSet gtid.Set) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if r.stage.Get() != int32(stageNew) {
+		return errors.Errorf("stage %s, expect %s", readerStage(r.stage), stageNew)
+	}
+
+	if gSet == nil {
+		return errors.NotValidf("nil GTID set")
+	}
+
+	streamer, err := r.syncer.StartSyncGTID(gSet.Origin())
+	if err != nil {
+		return errors.Annotatef(err, "start sync from GTID set %s", gSet)
+	}
+
+	r.streamer = streamer
+	r.stage.Set(int32(stagePrepared))
+	return nil
+}
+
+// Close implements Reader.Close.
+func (r *TCPReader) Close() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if r.stage.Get() == int32(stageClosed) {
+		return errors.New("already closed")
+	}
+
+	connID := r.syncer.LastConnectionID()
+	if connID > 0 {
+		dsn := fmt.Sprintf("%s:%s@tcp(%s:%d)/?charset=utf8mb4",
+			r.syncerCfg.User, r.syncerCfg.Password, r.syncerCfg.Host, r.syncerCfg.Port)
+		db, err := sql.Open("mysql", dsn)
+		if err != nil {
+			return errors.Annotatef(err, "open connection to the master %s:%d", r.syncerCfg.Host, r.syncerCfg.Port)
+		}
+		defer db.Close()
+		err = utils.KillConn(db, connID)
+		if err != nil {
+			return errors.Annotatef(err, "kill connection %d for master %s:%d", connID, r.syncerCfg.Host, r.syncerCfg.Port)
+		}
+	}
+
+	r.stage.Set(int32(stageClosed))
+	return nil
+}
+
+// GetEvent implements Reader.GetEvent.
+func (r *TCPReader) GetEvent(ctx context.Context) (*replication.BinlogEvent, error) {
+	if r.stage.Get() != int32(stagePrepared) {
+		return nil, errors.Errorf("stage %s, expect %s", readerStage(r.stage), stagePrepared)
+	}
+
+	return r.streamer.GetEvent(ctx)
+}
+
+// Status implements Reader.Status.
+func (r *TCPReader) Status() interface{} {
+	r.mu.Lock()
+	stage := r.stage
+	r.mu.Unlock()
```

Comment:
why only lock for stage getter? besides I find some redundant lock protection, such as 
```
r.mu.Lock()
defer r.mu.Unlock()
r.stage.Get()
```
Is there any consideration to do that?

Reply:
I change to use `stage = r.stage.Get` without lock.

We use `stage` both in comparing(L150) and recording(L154).
Saving `stage` in L148, so we can get the same value both in L150 and L154.